### PR TITLE
Keep parameter names for inlined queries

### LINKED ISSUE #23271: How to force reusing a parameter for inlined queries?
I've started to inline queries more often to reduce round-trips to the server for fetching IDs that later are used in other queries. 

This is partly due to the limitations of using `.Contains()` on an int-array. 

Instead I use something similar to the following (stupid) example, of course my real queries are different in terms of complexity. Still, this demonstrates the issue clearly.

```C#
                var offerId = 1;
                var inlineQuery = context.Offers.Where(o => o.Id == offerId);

                var firstQuery = context.OfferActions.Where(oa => inlineQuery.Contains(oa.Offer) && oa.Action == OfferActions.Established);
                var secondQuery = context.OfferActions.Where(oa => inlineQuery.Contains(oa.Offer) && oa.Action != OfferActions.Established);

                var query = firstQuery.Concat(secondQuery);
                await query.ToListAsync();
```

```
SELECT [o].[Id], [o].[Action], [o].[OfferId], [o].[TimeCreatedUtc]
      FROM [OfferActions] AS [o]
      INNER JOIN [Offers] AS [o0] ON [o].[OfferId] = [o0].[Id]
      WHERE EXISTS (
          SELECT 1
          FROM [Offers] AS [o1]
          WHERE ([o1].[Id] = @__offerId_0) AND ([o1].[Id] = [o0].[Id])) AND ([o].[Action] = 1)
      UNION ALL
      SELECT [o2].[Id], [o2].[Action], [o2].[OfferId], [o2].[TimeCreatedUtc]
      FROM [OfferActions] AS [o2]
      INNER JOIN [Offers] AS [o3] ON [o2].[OfferId] = [o3].[Id]
      WHERE EXISTS (
          SELECT 1
          FROM [Offers] AS [o4]
          WHERE ([o4].[Id] = @__offerId_1) AND ([o4].[Id] = [o3].[Id])) AND ([o2].[Action] <> 1)
```

As you can see in the above SQL output, I now have two parameters `__offerId_0` and `__offerId_1` even though I used the same local variable `offerId`.

Is there any way to force EF Core to reuse the parameters so that I only get a single parameter `__offerId `used in both subqueries?

In my real-world example, sometimes I end up with >10 parameters of the exact same value and I think this gives the SQL server a poor chance to optimize the query properly.

### Include provider and version information

EF Core version: 5.0 rc2
Database provider: Microsoft.EntityFrameworkCore.SqlServer

### PR DESCRIPTION
**Pull request by joakimriedel**
Fixes #23271 

New test `Query_reusing_parameter_with_inner_query_doesnt_declare_duplicate_parameter` added to verify solution.
Other tests modified since constants are now kept for lambdas and not reused from earlier top-level expressions.

All tests pass locally.

### CODE REVIEW

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 1]
```diff
@@ -40,6 +40,8 @@ public class ParameterExtractingExpressionVisitor : ExpressionVisitor
         private IDictionary<Expression, bool> _evaluatableExpressions;
         private IQueryProvider? _currentQueryProvider;
 
+        private int _reentryCount;
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:43]: No love for fields.
Add a protected method ExtractParameters (overload of current one) which takes a bool flag with default value of not clearing the dict. And the existing method would call into that with flag being true.

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs [hunk 2]
```diff
@@ -600,7 +600,7 @@ public override async Task Include_duplicate_collection(bool async)
     SELECT [c0].[CustomerID], [c0].[Address], [c0].[City], [c0].[CompanyName], [c0].[ContactName], [c0].[ContactTitle], [c0].[Country], [c0].[Fax], [c0].[Phone], [c0].[PostalCode], [c0].[Region]
     FROM [Customers] AS [c0]
     ORDER BY [c0].[CustomerID]
-    OFFSET 2 ROWS FETCH NEXT 2 ROWS ONLY
+    OFFSET @__p_0 ROWS FETCH NEXT @__p_0 ROWS ONLY
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: These are not really "same" in the tree sense so we need to do additional thing so that this test does not change.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 1 above)
Comment (joakimriedel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:43]: Could certainly be done with bool. But I still don't fully understand why dict needs to be cleared at all since no tests fail if I just never clear it. Is there some subtle edge case that do not have tests currently?

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs (hunk 2 above)
Comment (joakimriedel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: Spent some time digging deeper into this. It's interesting; constants within a `SelectMany` lambda will never be parameterized which means that if I change the test to use `Skip(3)` instead of `Skip(2)` it will become `OFFSET 3` (constant) instead of `OFFSET @__p_1` and `@__p_1=3`.  There's a comment in the code about this [here](https://github.com/dotnet/efcore/blob/a49074d816daeff6813931915df84c4e53f17d85/src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs#L555) where _inLambda will be true and then the code [here](https://github.com/dotnet/efcore/blob/a49074d816daeff6813931915df84c4e53f17d85/src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs#L540) will set generateParameter to false which is tested for [here](https://github.com/dotnet/efcore/blob/a49074d816daeff6813931915df84c4e53f17d85/src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs#L319) and the value 3 becomes constant. Should resolving this be part of this issue or open a new issue?

Comment (joakimriedel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: Thinking about this one step further, there are two different concerns here regardless of this PR.

1) since the constant 2 in first `Take` is automatically parameterized in this particular test, I believe constants in the `Skip` and second `Take` should also be parameterized for consistency

but I could also consider, and peerhaps even endorse,

2) no constants get ever parameterized - stay as constants in generated SQL. If you want a constant to be parameterized, declare a local variable `int takeCount = 2` and `Take(takeCount)`

how have you been discussing this internally ?

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs]: Memory footprint. This code is executed for every query (even ones used from the cache). It would be reclaimed at some point by GC but since we already know that we can erase it after the visitor ends, why not

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: Regardless of coming from variable or inline constant, when used outside of lambda, compiler evaluates value and puts it there. So even if you use `Take(takeCount)` in top level query, in expression it will be constant of 2 only and there will be no reference to takeCount variable. Hence Skip/Take try to generate parameter even if it is constant but that is only for top level. When it is inside lambda (like SelectMany), Skip/Take preserves whatever it was written with. So if it is coming from variable, we will see closure.takeCount member expression. But if the constant is inlined then we will see constant. Since we get actual information here, we need to generate constant if constant passed here.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 1 above)
Comment (joakimriedel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs]: Ah, thanks for clearing this up for me. Didn't occur to me that it was memory footprint, thought there were some edge case where the dict had to be cleared to force generating different parameter sets.

File: test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs (hunk 2 above)
Comment (joakimriedel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: Wow. Thanks for pointing this out, never knew the compiler acted like this. Makes me wonder - shouldn't all queries then go through lambda? A simple `context.Offers.Take(takeCount).SelectMany(t => t.Actions.Take(takeCount))` will produce two parameters `__p_0` for top-level and `__takeCount_1` for the lambda case. Tried `context.FromExpression(() => context.Offers.Take(takeCount).SelectMany(o => o.Actions.Take(takeCount)))` but it would still generate two parameters. How could a query run entirely in a lambda to get only `__takeCount_0` parameter?

Comment (joakimriedel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: Back to this PR, I have resolved this issue after our discussion above but now I get a different test failure. This time in `Constant_enum_with_same_underlying_value_as_previously_parameterized_int` 

where the following query

```c#
ss.Set<Gear>()
                    .OrderBy(g => g.Nickname)
                    .Take(1)
                    .Select(g => g.Rank & MilitaryRank.Private)
```

fails with the following assertion

```c#
@__p_0='1'

SELECT TOP(@__p_0) [g].[Rank] & @__p_0
FROM [Gears] AS [g]
ORDER BY [g].[Nickname]
```

since the fix I'm evaluating right now will render

```c#
@__p_0='1'

SELECT TOP(@__p_0) [g].[Rank] & 1
FROM [Gears] AS [g]
ORDER BY [g].[Nickname]
```

since the `MilitaryRank.Private` is a constant within a lambda. According to our discussion above, I think this test is wrong and should not reuse the parameter for a supplied constant like that. OK to change the test assertion?

Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/NorthwindIncludeQuerySqlServerTest.cs:603]: Seems like correct change.

**Comment by joakimriedel** [comment]
Ok, this is ready for review now. Fixes three separate issues, see first comment above.

**Comment by joakimriedel** [comment]
All good @smitpatel ? Let me know if there is more work to do here.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 3]
```diff
@@ -77,6 +77,18 @@ public class ParameterExtractingExpressionVisitor : ExpressionVisitor
...
+            return ExtractParameters(expression, clearEvaluatedValues: true);
+        }
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        protected virtual Expression ExtractParameters([NotNull] Expression expression,
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:90]: Apologies for earlier comment. We can just make this private. No one would be using this anyway

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 4]
```diff
@@ -302,11 +317,7 @@ private Expression Evaluate(Expression expression, bool generateParameter)
 
                 if (!generateParameter)
                 {
-                    var constantValue = GenerateConstantExpression(parameterValue, expression.Type);
-
-                    _evaluatedValues.Add(expression, constantValue);
-
-                    return constantValue;
+                    return GenerateConstantExpression(parameterValue, expression.Type);
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:320]: We still want to use cached value even for constant. Evaluating non-trivial expression is expensive hence we have caching. But here is the catch, 
- We want to look up only using expression
  - If the earlier evaluated expression is constant then we can extract value from the constantExpression
  - If the earlier evaluated expression is parameter than we can get the value from _parameterValues
- When earlier evaluated value is constant we still want to know the candidate parameterName generated for it so we can follow the convention for naming when generating parameter at later stage.
When earlier evaluated value is parameter, we should also cache the constant value in some way so we don't have to look up _parameterValue always.

One idea I can think of is that use a struct for cached value.
So Dictionary<Expression, CachedValue>(ExpressionEqualityComparer.Instance)
CachedValue struct can hold the evaluated value itself
+ parameterName
and optionally parameterExpression/ConstantExpression as they are generated for use.
That way we can easily access and get previously used expression at the same time generate (and store) the other as needed.

File: test/EFCore.Specification.Tests/BuiltInDataTypesTestBase.cs [hunk 5]
```diff
@@ -2011,61 +2013,73 @@ public virtual void Can_compare_enum_to_parameter()
...
-            Assert.Equal(expected.Short, actual.Short);
-            Assert.Equal(expected.Ushort, actual.Ushort);
-            Assert.Equal(expected.Int, actual.Int);
-            Assert.Equal(expected.Uint, actual.Uint);
-            Assert.Equal(expected.Long, actual.Long);
-            Assert.Equal(expected.Ulong, actual.Ulong);
-            Assert.Equal(expected.Decimal, actual.Decimal);
-            Assert.Equal(expected.Char, actual.Char);
+            var oldCulture = Thread.CurrentThread.CurrentCulture;
+            try
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/BuiltInDataTypesTestBase.cs:2018]: Remove this try catch.

We have #14912 to deal with culture issue since there are more tests which fails in other cultures. This test will fail in your local build but it will pass in PR check.

If you really want to make this test pass locally too 😉 then just annotate the test with UseCultureAttribute with en-us

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 3 above)
Comment (joakimriedel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs]: No worries, reverted.

File: test/EFCore.Specification.Tests/BuiltInDataTypesTestBase.cs (hunk 5 above)
Comment (joakimriedel) [reply test/EFCore.Specification.Tests/BuiltInDataTypesTestBase.cs:2018]: Actually all tests pass, except this one. It's not related to #14912 but #22901 since it's a problem with decimal point vs decimal comma (like here in Sweden) when converting to string client evaluated vs InMemory provider.

UseCultureAttribute looks handy. Is there any similar attribute for Invariant culture? That's why I added the try finally to ensure the invariant culture does not stick after test failure on the thread.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 4 above)
Comment (joakimriedel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:313]: I actually sampled quite a few test cases for this solution, but could not find a single one where the expression would be calculated more than once per expression. That's why I chose the easy solution to simply remove the caching whenever a parameter was not generated. If you could point me to a test where it would actually benefit from the cache by requesting the expression to be translated twice (hitting the cache), I could experiment with a caching solution. I think the easiest way then would be to change the key of the existing dictionary to a tuple (Expression, bool /generateParameter/) which would automatically handle the generate vs not-generate cases.

File: test/EFCore.Specification.Tests/BuiltInDataTypesTestBase.cs (hunk 5 above)
Comment (smitpatel) [reply test/EFCore.Specification.Tests/BuiltInDataTypesTestBase.cs:2018]: It is #14192 in a way that Swedish culture and en-us only mismatch for decimal point/comma. But other cultures varies for other formatting also. So this would pass the test for swedish culture but all tests still does not pass for all cultures. So when we fix it, it will also fix this case.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 4 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:313]: So for most cases, constant would be created only when it is a constant in the tree. Even NewExpression preserves itself. So what are the constant we have at hand other than primitive types.
But then we have query filters which explicitly inline the variables captured as constants, as long as they are not rooted at DbContext. That is the scenario where it would be helpful.
For test a query filter which is doing
`HasQueryFilter(e => e.Id == (a + b + c + d + e))`
would need to generate a constant and repeat it multiple times in the filter in some way. Artificial case in a way but this parameter extractor is the most naughtiest part of the pipeline.
I suggested a struct approach since we don't even want to evaluate it twice (once for parameter and once for constant).

**Comment by smitpatel** [comment]
Requesting review from @roji / @AndriySvyryd to look over the simple caching mechanism.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 6]
```diff
@@ -76,6 +76,11 @@ public class ParameterExtractingExpressionVisitor : ExpressionVisitor
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
         public virtual Expression ExtractParameters([NotNull] Expression expression)
+        {
+            return ExtractParameters(expression, clearEvaluatedValues: true);
+        }
+
+        private Expression ExtractParameters([NotNull] Expression expression, bool clearEvaluatedValues = false)
```
Comment (roji) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:83]: We should remove the default value for `clearEvaluatedValues` - it's ambiguous with the other overload.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 7]
```diff
@@ -280,16 +288,29 @@ private static Expression GenerateConstantExpression(object? value, Type returnT
 
         private Expression Evaluate(Expression expression, bool generateParameter)
         {
+            object? parameterValue;
+            string? parameterName;
```
Comment (roji) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:292]: Might be good to look at nullability here - under what circumstances is parameterName null? GetValue seems to return with parameterName null only when its expression parameter is null, but I'm not sure that's ever possible...

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 8]
```diff
@@ -280,16 +288,29 @@ private static Expression GenerateConstantExpression(object? value, Type returnT
...
+            {
+                parameterValue = GetValue(expression, out parameterName);
+                cachedValue = new EvaluatedValues { CandidateParameterName = parameterName, Value = parameterValue };
+                _evaluatedValues[expression] = cachedValue;
+            }
 
             if (parameterValue is IQueryable innerQueryable)
             {
-                return ExtractParameters(innerQueryable.Expression);
+                return ExtractParameters(innerQueryable.Expression, clearEvaluatedValues: false);
```
Comment (roji) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:313]: There's another ExtractParameter just below on line 20 - should it also pass `clearEvaluatedValues: false`?

**Review by roji — COMMENTED** [review]
@smitpatel gave it a look, but probably good to have @AndriySvyryd eyes on it too.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 6 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:83]: private function? does it matter?

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 7 above)
Comment (smitpatel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:292]: Parameter name is populated only when there is memberInfo or a method. Something as simple as a constant also returns null parameter name. (which is converted to p later but candidate name is still null)

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs (hunk 6 above)
Comment (roji) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:83]: Just remove the default value maybe? It doesn't really matter much, but this does generate a warning in some IDEs etc.

File: src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs [hunk 9]
```diff
@@ -645,5 +666,13 @@ private static bool IsQueryableMethod(Expression expression)
                 => expression is MethodCallExpression methodCallExpression
                     && methodCallExpression.Method.DeclaringType == typeof(Queryable);
         }
+
+        private sealed class EvaluatedValues
```
Comment (AndriySvyryd) [inline src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:670]: Would this ever be used to create a cache key?

Comment (smitpatel) [reply src/EFCore/Query/Internal/ParameterExtractingExpressionVisitor.cs:670]: Nope. This is cached value. Key is actual expression.

**Comment by AndriySvyryd** [comment]
:shipit:

**Review by smitpatel — APPROVED** [review]
